### ISSUE TITLE
Crash on Android when desarializing Task objects from JSON: FlutterError: type 'double' is not a subtype of type 'int'

### ISSUE THREAD

**Issue by snipd-mikel** [781flyingdutchman/background_downloader#70]
Hi,

We have been seeing some crash reports on Crashlytics when deserializing Task objects:

```
FlutterError: type 'double' is not a subtype of type 'int'
       at new Task.fromJsonMap(models.dart:406)
       at new DownloadTask.fromJsonMap(models.dart:509)
       at new Task.createFromJsonMap(models.dart:362)
       at BaseDownloader.retrieveLocallyStoredData(base_downloader.dart:105)
       at MobileDownloadServiceImpl.initialize(download.service.new.dart:92)
```

The `Task.fromJsonMap(models.dart:406)` line is the [following](https://github.com/781flyingdutchman/background_downloader/blob/59a0236722a2f6f835440a5e9a35e8bbb61bf6f2/lib/src/models.dart#L406):

```dart
 baseDirectory = BaseDirectory.values[jsonMap['baseDirectory'] ?? 0],
```

It seems that `jsonMap['baseDirectory']` is decoded as `double` instead of `int` on Android.

You use `Gson` to encode objects to json on the Android side. It seems that Gson defaults to using double for numbers when decoding JSON according to [this](https://github.com/google/gson/blob/main/Troubleshooting.md#-integral-json-number-is-parsed-as-double). I'm not sure if this is also happening when encoding the JSON to string.

The method that fails is the one that calls `popLocalStorage` on the native side. That method is directly returning the string on the SharedPreferences. 

One possible solution is to cast numbers to `num` on the dart side and then convert them to `int` or `double`. I'll create a PR with this.

**Comment by 781flyingdutchman** [comment]
Thanks for calling this out. I've seen this issue pop up related to the `TaskProgressUpdate` (and updated the code there) but had not seen this happen related to `Task`. I'm a little suspicious, as the task moves to and from JSON all the time, so unclear why this becomes an issue only when executed as part of `popLocalStorage` so let me look into this a bit. Really appreciate your pull request - just want to make sure that we address this issue everywhere it may surface.

**Comment by 781flyingdutchman** [comment]
Implemented in V7.4.1 (and V6.2.1 if you're on Dart 2). Please let me know if you still see Crashlytics crashes of this type for that version - otherwise I assume the issue has been fixed

**Comment by snipd-mikel** [comment]
Thanks for the reply and for merging the fix.

>  I'm a little suspicious, as the task moves to and from JSON all the time, so unclear why this becomes an issue only when executed as part of popLocalStorage so let me look into this a bit. Really appreciate your pull request - just want to make sure that we address this issue everywhere it may surface.

I'm also suspicious as why the issue it's happening at all. Looking at the path that JSON takes, I don't see how the Int can become a double. I was not able to reproduce so I only had the crashes from Crashlytics. I think my fix is just a patch, but it doesn't address the root of the cause.

> Implemented in V7.4.1 (and V6.2.1 if you're on Dart 2). Please let me know if you still see Crashlytics crashes of this type for that version - otherwise I assume the issue has been fixed

Thanks for merging, I will now release a new update and keep an eye on the issue. Will let you know next week :)

**Comment by snipd-mikel** [comment]
I just got new crashes from Crashlytics:

```
new Request.fromJsonMap (models.dart:217)
new Task.fromJsonMap (models.dart:413)
new DownloadTask.fromJsonMap (models.dart:510)
new Task.createFromJsonMap (models.dart:362)
BaseDownloader.retrieveLocallyStoredData (base_downloader.dart:96)
MobileDownloadServiceImpl.initialize (download.service.new.dart:93)
Future.wait.<fn> (dart:async)
.main (main.dart:98)
```

Seems that the fix worked but now it's crashing at a later point, for the same reason. I have created a new PR to address this. I will however try to investigate this further so we can fix the underlying issue, as this is very error prone and likely to happen again every time a new int field is added to those classes.

**Comment by 781flyingdutchman** [comment]
Hi, something is wrong here though: the current main branch (V7.4.1 with the fix I published) already has the num toInt solution. Your latest PR appears to be based off an older branch - not the latest package branch - and I have already made those changes in the published version. Not sure where there misalignment is - can you try with the published latest version and let me know if you get this error?

**Comment by snipd-mikel** [comment]
Aaa sorry! I didn't see that. I released yesterday from my fork, before you merged and released 7.4.1. Then I supposed this issue is fixed. You can close the issue and PR then

**Comment by 781flyingdutchman** [comment]
No problem - please try with the published version and let me know if you still run into issues.
